File: Prácticas/P3/Java/civitas/src/civitas/Sorpresa.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;

/**
 *
 * @author Jose Padial
 */
public class Sorpresa {
    private String texto;
    private int valor;
    private MazoSorpresas mazo;
    private TipoSorpresa tipo;
    private Tablero tablero;
    private Diario diario = Diario.getInstance();
    
    Sorpresa(TipoSorpresa tipo, Tablero tablero){
        init();
        this.tipo = tipo;
        this.tablero = tablero;
        this.texto = "Carcel";
    }
    
    Sorpresa(TipoSorpresa tipo, Tablero tablero, int valor, String texto){
        init();
        this.tipo = tipo;
        this.tablero = tablero;
        this.valor = valor;
        this.texto = texto;
    }
    
    Sorpresa(TipoSorpresa tipo, int valor, String texto){
        init();
        this.tipo = tipo;
        this.valor = valor;
        this.texto = texto;
    }
    
    Sorpresa(TipoSorpresa tipo, MazoSorpresas mazo){
        init();
        this.tipo = tipo;
        this.mazo = mazo;
        this.texto = "Evita la carcel";
    }
    
    void aplicarAJugador(int actual, ArrayList<Jugador> todos){
        switch(tipo){
            case IRCASILLA:
                aplicarAJugador_irACasilla(actual, todos);
                break;
            case IRCARCEL:
                aplicarAJugador_irCarcel(actual, todos);
                break;
            case PAGARCOBRAR:
                aplicarAJugador_pagarCobrar(actual, todos);
                break;  
            case PORCASAHOTEL:
                aplicarAJugador_porCasaHotel(actual, todos);
                break;  
            case PORJUGADOR:
                aplicarAJugador_porJugador(actual, todos);
                break;    
            case SALIRCARCEL:
                aplicarAJugador_salirCarcel(actual, todos);
                break;
        }
    }
    
    private void aplicarAJugador_irACasilla(int actual, ArrayList<Jugador> todos){
        if(juagadorCorrecto(actual, todos)){
            informe(actual, todos);
            int casillaActual = todos.get(actual).getNumCasillaActual();
            int tirada = tablero.calcularTirada(casillaActual, valor);
            int nuevaPos = tablero.nuevaPosicion(casillaActual, tirada);
            todos.get(actual).moverACasilla(nuevaPos);
            tablero.getCasilla(nuevaPos).recibeJugador(actual, todos);
        }
    }
    
    private void aplicarAJugador_irCarcel(int actual, ArrayList<Jugador> todos){
        if(juagadorCorrecto(actual, todos)){
            informe(actual, todos);
            todos.get(actual).encarcelar(tablero.getCarcel());
        }
    }
    
    private void aplicarAJugador_pagarCobrar(int actual, ArrayList<Jugador> todos){
        if(juagadorCorrecto(actual, todos)){
            informe(actual, todos);
            todos.get(actual).modificarSaldo(valor);
        }
    }
    
    private void aplicarAJugador_porCasaHotel(int actual, ArrayList<Jugador> todos){
        if(juagadorCorrecto(actual, todos)){
            informe(actual, todos);
            todos.get(actual).modificarSaldo(valor*todos.get(actual).cantidadCasasHoteles());
        }
    }
    
    private void aplicarAJugador_porJugador(int actual, ArrayList<Jugador> todos){
        if(juagadorCorrecto(actual, todos)){
            informe(actual, todos);
            Sorpresa aux = new Sorpresa(TipoSorpresa.PAGARCOBRAR, valor*-1, "aux");
            Sorpresa aux2 = new Sorpresa(TipoSorpresa.PAGARCOBRAR, valor*(todos.size()-1), "aux2");
            for(int i=0; i<todos.size(); i++){
                if(i != actual){
                    aux.aplicarAJugador(i, todos);
                }
                else
                    aux2.aplicarAJugador(actual, todos);
            }
        }
    }
    
    private void aplicarAJugador_salirCarcel(int actual, ArrayList<Jugador> todos){
        boolean puede = false;
        if(juagadorCorrecto(actual, todos)){
            informe(actual, todos);
            for(Jugador j:todos){
                if(j.tieneSalvoconducto() == true && j != todos.get(actual))
                    puede = true;
            }
            if(!puede){
                todos.get(actual).obtenerSalvoconducto(this);
                salirDelMazo();
            }
        }
    }
    
    private void informe(int actual, ArrayList<Jugador> todos){
        diario.ocurreEvento("Se aplica la sorpresa " + texto + " a " + todos.get(actual).getNombre());
    }
    
    private void init(){
        this.valor = -1;
        this.mazo = new MazoSorpresas();
        this.tablero = new Tablero(10);
        this.texto = "";
    }
    
    public boolean juagadorCorrecto(int actual, ArrayList<Jugador> todos){
        return todos.size() > actual;
    }
    
    void salirDelMazo(){
        if(tipo == TipoSorpresa.SALIRCARCEL)
            mazo.inhabilitarCartaEspecial(this);
    }
    
    void usadas(){
        if(tipo == TipoSorpresa.SALIRCARCEL)
            mazo.habilitarCartaEspecial(this);
    }

    @Override
    public String toString() {
        return "Sorpresa{" + "texto=" + texto + ", valor=" + valor + ", mazo=" + mazo + ", tipo=" + tipo + ", tablero=" + tablero + '}';
    }
}

File: Prácticas/P3/Ruby/civitas/pkg/civitas-0.0.1/lib/sorpresa.rb
# encoding: utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative "tipo_sorpresa.rb"
require_relative "mazo_sorpresas.rb"
require_relative "tablero.rb"
require_relative "diario.rb"
require_relative "jugador.rb"
require_relative "casilla.rb"

module Civitas
  class Sorpresa
    def self.ini_carcel(tipo,tablero)
      self.new(tipo)
      @tablero = tablero
      @texto = "Carcel"
    end
    
    def self.ini_mov_otra(tipo,tablero,valor,texto)
      self.new(tipo)
      @tablero = tablero
      @valor = valor
      @texto = texto
    end
    
    def self.ini_exit(tipo,mazo)
      self.new(tipo)
      @mazo = mazo
      @texto = "Evita la carcel"
    end
    
    def self.ini_resto(tipo,valor,texto)
      self.new(tipo)
      @valor = valor
      @texto = texto
    end
    
    def initialize(tipo)
      init()
      @tipo = tipo
      @diario = Diario.instance
    end
    
    # private_class_method :new
     
    private
    
    def aplicar_a_jugador_ir_a_casilla(actual,todos)
      if(jugador_correcto(actual,todos))
        informe(actual,todos)
        casilla_actual = todos.at(actual).num_casilla_actual
        tirada = @tablero.calcular_tirada(casilla_actual, @valor)
        nueva_pos = @tablero.nueva_posicion(casilla_actual, tirada)
        todos.at(actual).mover_a_casilla(nueva_pos)
        @tablero.get_casilla(nueva_pos).recibe_jugador(actual,todos)
      end
    end
    
    def aplicar_a_jugador_ir_carcel(actual,todos)
      if(jugador_correcto(actual,todos))
        informe(actual,todos)
        todos.at(actual).encarcelar(@tablero.num_casilla_carcel)
      end
    end
    
    def aplicar_a_jugador_pagar_cobrar(actual,todos)
      if(jugador_correcto(actual,todos))
        informe(actual,todos)
        todos.at(actual).modificar_saldo(@valor)
      end
    end
    
    def aplicar_a_jugador_por_casa_hotel(actual,todos)
      if(jugador_correcto(actual,todos))
        informe(actual,todos)
        todos.at(actual).modificar_saldo(@valor*todos.at(actual).cantidad_casas_hoteles())
      end
    end
    
    def aplicar_a_jugador_por_jugador(actual,todos)
      if(jugador_correcto(actual,todos))
        informe(actual,todos)
        aux = Sorepresa.ini_resto(TipoSorpresa::PAGARCOBRAR,valor*-1,"aux")
        aux2 = Sorepresa.ini_resto(TipoSorpresa::PAGARCOBRAR,valor*(todos.lenght),"aux2")
        for i in 1..todos.lenght
          if(i != actual)
            aux.aplicar_a_jugador(i,todos)
          else
            aux2.aplicar_a_jugador(actual,todos)
          end
        end
      end
    end
    
    def aplicar_a_jugador_salir_carcel(actual,todos)
      salida = false
      if(jugador_correcto(actual,todos))
        informe(actual,todos)
        for i in 1..todos.lenght
          if(i.tiene_salvoconducto() == true && i != actual)
            salida = true
          end
        end
        if(!salida)
          todos.at(actual).obtener_salvoconducto(self)
          salir_del_mazo()
        end
      end
      salida
    end
    
    def informe(actual,todos)
      diario.ocurre_evento("Se aplica la sorpresa " + @texto + " a " + todos.at(actual).nombre)
    end
    
    def init()
      @valor = -1
      @mazo = MazoSorpresas.new()
      @tablero = Tablero.new(10)
      @texto = ""
    end
    
    public
     
    def apliacar_a_jugador(actual,todos)
      case @tipo
        when TipoSorpresa::IRCASILLA
          aplicar_a_jugador_ir_a_casilla(actual,todos)
        when TipoSorpresa::IRCARCEL
          aplicar_a_jugador_ir_carcel(actual,todos)
        when TipoSorpresa::PAGARCOBRAR
          aplicar_a_jugador_pagar_cobrar(actual,todos)
        when TipoSorpresa::PORCASAHOTEL
          aplicar_a_jugador_por_casa_hotel(actual,todos)
        when TipoSorpresa::PORJUGADOR
          aplicar_a_jugador_por_jugador(actual,todos)
        when TipoSorpresa::SALIRCARCEL
          aplicar_a_jugador_salir_carcel(actual,todos)
      end
    end
    
    def jugador_correcto(actual,todos)
      todos.lenght > actual
    end
    
    def salir_del_mazo()
      if(@tipo == TipoSorpresa::SALIRCARCEL)
        @mazo.inhabilitar_carta_especial(self)
      end
    end
    
    def usadas()
      if(@tipo == TipoSorpresa::SALIRCARCEL)
        @mazo.habilitar_carta_especial(self)
      end
    end
    
    def to_s
      "Sorpresa{texto=#{@texto}, valor=#{@valor}, tipo=#{@tipo}}"
    end
    
  end
end
